Add unique index to DataFlagOpinions

nritsche opened the pull request:
Adds a unique index combining the columns `type`, `user`, `start_time`, `finish_time`.

nritsche committed:
style: run black 20.8b1

nritsche committed:
feat(DataFlagOpinion): add unique index

nritsche commented:
This should really include metadata. But we would have to make that fixed length to be accepted as part of the index and we are not using that metadata yet, anyways.

jrs65 commented:
Can you explain again why you'd want to index on metadata? I would have thought indexing on `type`, `user` `start/finish_time` and `rev` would be sufficient? i.e. a user can only submit one opinion per day per revision of a piece of data.

anjakefala commented on chimedb/dataflag/orm.py:
```diff
@@ -261,6 +261,12 @@ class DataFlagOpinion(DataSubset):
+    class Meta:
+        indexes = (
+            # create a unique index
+            (("type", "user", "start_time", "finish_time"), True),
```
What will the conflict resolution strategy be?

nritsche replied:
in bondia the user is warned that they already inserted an opinion, if they click anyways, the `decision` field is updated.

jrs65 commented on chimedb/dataflag/orm.py:
```diff
@@ -261,6 +261,12 @@ class DataFlagOpinion(DataSubset):
+    class Meta:
+        indexes = (
+            # create a unique index
+            (("type", "user", "start_time", "finish_time"), True),
```
I think you should also index on `revision` in here.

nritsche replied:
yes

nritsche commented:
> Can you explain again why you'd want to index on metadata? I would have thought indexing on `type`, `user` `start/finish_time` and `rev` would be sufficient? i.e. a user can only submit one opinion per day per revision of a piece of data.

Yes, but the idea of metadata was to later add frequency and input ranges here (see https://github.com/chime-experiment/chimedb_dataflag/blob/f3e8a96c5259bb874bd727b4cb17616baacdfa97/chimedb/dataflag/orm.py#L235).

nritsche committed:
feat(orm): opinions are fixed to LSDs

- remove start and finish time from opinion
- remove metadata field from opinion
- add lsd field to opinion
- judge translates from lsd to start and finish times when creating
flags

nritsche committed:
fix(vote): only create flags for bad data

create vote anyways

nritsche committed:
refactor(DataFlag): re-simplify by removing baseclass

nritsche committed:
ci: move from travis to gihub

nritsche commented:
I think I need https://github.com/chime-experiment/chimedb/issues/18 to find out what's wrong here.

nritsche committed:
feat(opinion): add optional notes field for comments

jrs65 commented on chimedb/dataflag/orm.py:
```diff
@@ -260,6 +298,14 @@ class DataFlagOpinion(DataSubset):
     client = pw.ForeignKeyField(DataFlagClient)
     revision = pw.ForeignKeyField(DataRevision, backref="opinions")
+    lsd = pw.IntegerField()
+    note = pw.TextField(null=True)
```
Everywhere else we call this field `notes`.
